File: 2022/matlab/aoc3.m
tic
input_path = fullfile(get_input_root,'day_3.txt');

x = read_txt(input_path);
lines = strsplit(x);

letters_upper = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
letters_lower = lower(letters);
%part 1
compartments = cell(numel(lines),2);
intersection_chars = cell(size(lines));

score = nan(size(lines));
for i1 = 1:numel(lines)
    line_i1 = lines{i1};
    n_char = numel(line_i1)/2;
    %n_char
    compartments{i1,1} = line_i1(1:n_char);
    compartments{i1,2} = line_i1(n_char+1:end);
    
    intersection_chars{i1} = intersect(compartments{i1,1},compartments{i1,2});
    i_lower = find(letters_lower == intersection_chars{i1});
    i_upper = find(letters_upper == intersection_chars{i1});
    if ~isempty(i_lower)
        score(i1) = i_lower;
    elseif ~isempty(i_upper)
        score(i1) = i_upper+26;
    end
end
sum(score)

%% part 2 
intersection_chars = cell(size(lines));

group_inds = [1:3:numel(lines)-2];
score = nan(size(group_inds));
for i1 = 1:numel(group_inds)
    group_i1 = lines(group_inds(i1):group_inds(i1)+2);
    c_intersect = intersect(intersect(group_i1{1},group_i1{2}),group_i1{3});
    i_lower = find(letters_lower == c_intersect);
    
    i_upper = find(letters_upper == c_intersect);
    if ~isempty(i_lower)
        score(i1) = i_lower;
    elseif ~isempty(i_upper)
        score(i1) = i_upper+26;
    end
%         
end
sum(score)

File: 2022/python/aoc3.py
input_path = '../inputs/day_3.txt'
lines = open(input_path).read().strip().split('\n')

def get_points(x):
    letters = 'abcdefghijklmnopqrstuvwxyz' 
    letters = letters + letters.upper()
    return letters.find(x) + 1
    
points = 0   

#part 1
for i,l in enumerate(lines):
    nc = int(len(l)/2)
    c1 = set(l[0:nc])
    c2 = set(l[nc:])
    c_intersect = list(set.intersection(c1,c2))[0]
    points += get_points(c_intersect)

print(points)

#part2 
points = 0   
for i in range(0,len(lines),3):
    group_i = lines[i:i+3]
    group_sets = list(map(set,group_i))
    c_intersect = list(set.intersection(*group_sets))[0]
    points += get_points(c_intersect)
    
print(points)